freebsd4/usr.sbin/lpr/lpr/printcap.5
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)printcap.5	8.2 (Berkeley) 12/11/93
.\" $FreeBSD: src/usr.sbin/lpr/lpr/printcap.5,v 1.16 1999/08/28 01:16:55 peter Exp $
.\"
.Dd October 11, 2000
.Dt PRINTCAP 5
.Os BSD 4.2
.Sh NAME
.Nm printcap
.Nd printer capability data base
.Sh SYNOPSIS
.Nm
.Sh DESCRIPTION
The
.Nm Printcap
function
is a simplified version of the
.Xr termcap 5
data base
used to describe line printers.  The spooling system accesses the
.Nm
file every time it is used, allowing dynamic
addition and deletion of printers.  Each entry in the data base
is used to describe one printer.  This data base may not be
substituted for, as is possible for 
.Xr termcap 5 ,
because it may allow accounting to be bypassed.
.Pp
The default printer is normally 
.Em lp ,
though the environment variable
.Ev PRINTER
may be used to override this.  Each spooling utility supports an option,
.Fl P Ar printer ,
to allow explicit naming of a destination printer.
.Pp
Refer to the
.%T "4.3 BSD Line Printer Spooler Manual"
for a complete discussion on how to setup the database for a given printer.
.Sh CAPABILITIES
Refer to
.Xr termcap 5
for a description of the file layout.
.Bl -column Namexxx Typexx "/var/spool/lpdxxxxx"
.Sy Name	Type	Default	Description
.It "af	str" Ta Dv NULL Ta No "name of accounting file"
.It "br	num	none	if lp is a tty, set the baud rate"
.Pf ( Xr ioctl 2
call)
.It "cf	str" Ta Dv NULL Ta No "cifplot data filter"
.It "ct	num	120	TCP connection timeout in seconds"
.It "df	str" Ta Dv NULL Ta No "tex data filter"
.Pf ( Tn DVI
format)
.It "ff	str" Ta So Li \ef Sc Ta No "string to send for a form feed"
.It "fo	bool	false	print a form feed when device is opened"
.It "gf	str" Ta Dv NULL Ta No "graph data filter"
.Pf ( Xr plot 3
format
.It "hl	bool	false	print the burst header page last"
.It "ic	bool	false	driver supports (non standard) ioctl to indent printout"
.It "if	str" Ta Dv NULL Ta No "name of text filter which does accounting"
.It "lf	str" Ta Pa /dev/console Ta No "error logging file name"
.It "lo	str" Ta Pa lock Ta No "name of lock file"
.It "lp	str" Ta Pa /dev/lp Ta No "device name to open for output"
.It "ms	str" Ta Dv NULL Ta No "if lp is a tty, a comma-separated, "
.Xr stty 1
-like list describing the tty modes
.It "mx	num	1000	maximum file size (in"
.Dv BUFSIZ
blocks), zero = unlimited
.It "nd	str" Ta Dv NULL Ta No "next directory for list of queues (unimplemented)"
.It "nf	str" Ta Dv NULL Ta No "ditroff data filter (device independent troff)"
.It "of	str" Ta Dv NULL Ta No "name of output filtering program"
.It "pc	num	200	price per foot or page in hundredths of cents"
.It "pl	num	66	page length (in lines)"
.It "pw	num	132	page width (in characters)"
.It "px	num	0	page width in pixels (horizontal)"
.It "py	num	0	page length in pixels (vertical)"
.It "rf	str" Ta Dv NULL Ta No "filter for printing"
.Tn FORTRAN
style text files
.It "rg	str" Ta Dv NULL Ta No "restricted group. Only members of group allowed access"
.It "rm	str" Ta Dv NULL Ta No "machine name for remote printer"
.It "rp	str	``lp''	remote printer name argument"
.It "rs	bool	false	restrict remote users to those with local accounts"
.It "rw	bool	false	open the printer device for reading and writing"
.It "sb	bool	false	short banner (one line only)"
.It "sc	bool	false	suppress multiple copies"
.It "sd	str" Ta Pa /var/spool/lpd Ta No "spool directory"
.It "sf	bool	false	suppress form feeds"
.It "sh	bool	false	suppress printing of burst page header"
.It "sr	str" Ta Dv NULL Ta No "file name to hold statistics of each datafile as it is received"
.It "ss	str" Ta Dv NULL Ta No "file name to hold statistics of each datafile as it is sent"
.It "st	str" Ta Pa status Ta No "status file name"
.It "tf	str" Ta Dv NULL Ta No "troff data filter (cat phototypesetter)"
.It "tr	str" Ta Dv NULL Ta No "trailer string to print when queue empties"
.It "vf	str" Ta Dv NULL Ta No "raster image filter"
.El
.Pp
Each two-letter capability has a human-readable alternate name.
.Bl -column "Short form" "Long form"
.Sy "Short form	Long form"
.It "af	acct.file"
.It "br	tty.rate"
.It "cf	filt.cifplot"
.It "ct	remote.timeout"
.It "df	filt.dvi"
.It "du	daemon.user"
.It "ff	job.formfeed"
.It "fo	job.topofform"
.It "gf	filt.plot"
.It "hl	banner.last"
.It "if	filt.input"
.It "lf	spool.log"
.It "lo	spool.lock"
.It "lp	tty.device"
.It "mc	max.copies"
.It "ms	tty.mode"
.It "mx	max.blocks"
.It "nf	filt.ditroff"
.It "of	filt.output"
.It "pc	acct.price"
.It "pl	page.length"
.It "pw	page.width"
.It "px	page.pwidth"
.It "py	page.plength"
.It "rf	filt.fortran"
.It "rg	daemon.restrictgrp"
.It "rm	remote.host"
.It "rp	remote.queue"
.It "rs	daemon.restricted"
.It "rw	tty.rw"
.It "sb	banner.short"
.It "sc	job.no_copies"
.It "sd	spool.dir"
.It "sf	job.no_formfeed"
.It "sh	banner.disable"
.It "sr	stat.recv"
.It "ss	stat.send"
.It "st	spool.status"
.It "tr	job.trailer"
.It "vf	filt.raster"
.El
.Pp
If the local line printer driver supports indentation, the daemon
must understand how to invoke it.
.Sh FILTERS
The
.Xr lpd 8
daemon creates a pipeline of
.Em filters
to process files for various printer types.
The filters selected depend on the flags passed to
.Xr lpr 1 .
The pipeline set up is:
.Bd -literal -offset indent
p	pr | if	regular text + pr(1)
none	if	regular text
c	cf	cifplot
d	df	DVI (tex)
g	gf	plot(3)
n	nf	ditroff
f	rf	Fortran
t	tf	troff
v	vf	raster image
.Ed
.Pp
The
.Sy if
filter is invoked with arguments:
.Bd -filled -offset indent
.Cm if
.Op Fl c
.Fl w Ns Ar width
.Fl l Ns Ar length
.Fl i Ns Ar indent
.Fl n Ar login
.Fl h Ar host acct-file
.Ed
.Pp
The
.Fl c
flag is passed only if the
.Fl l
flag (pass control characters literally)
is specified to
.Xr lpr 1 .
The
.Ar Width
function
and
.Ar length
specify the page width and length
(from
.Cm pw
and
.Cm pl
respectively) in characters.
The
.Fl n
and
.Fl h
parameters specify the login name and host name of the owner
of the job respectively.
The
.Ar Acct-file
function
is passed from the
.Cm af
.Nm
entry.
.Pp
If no
.Cm if
is specified,
.Cm of
is used instead,
with the distinction that
.Cm of
is opened only once,
while
.Cm if
is opened for every individual job.
Thus,
.Cm if
is better suited to performing accounting.
The
.Cm of
is only given the
.Ar width
and
.Ar length
flags.
.Pp
All other filters are called as:
.Bd -filled -offset indent
.Nm filter
.Fl x Ns Ar width
.Fl y Ns Ar length
.Fl n Ar login
.Fl h Ar host acct-file
.Ed
.Pp
where
.Ar width
and
.Ar length
are represented in pixels,
specified by the
.Cm px
and
.Cm py
entries respectively.
.Pp
All filters take
.Em stdin
as the file,
.Em stdout
as the printer,
may log either to
.Em stderr
or using
.Xr syslog 3 ,
and must not ignore
.Dv SIGINT . 
.Sh REMOTE PRINTING
When printing to a remote printer using
.Cm rm ,
it is possible to use either
.Cm if
or
.Cm of . 
If both are specified,
.Cm of
is ignored.  Both filters behave the same except that they are passed
different arguments as above.  Specifically, the output filter is
terminated and restarted for each file transmitted.  This is necessary
in order to pass the resulting size to the remote
.Xr lpd 8 .
.Pp
If the
.Fl p
flag was passed to
.Xr lpr 1 ,
.Xr pr 1
is not executed locally, but is requested of the remote
.Xr lpd 8 .
Any input filtering via
.Cm if
will therefore happen before
.Xr pr 1
is executed rather than afterwards.
.Pp
If
.Cm lp
is specified as
.Em port Ns No @ Ns Em machine
(and
.Cm rm
is not in use), printing will be send directly to the given
.Em port
on the given
.Em machine .
.Sh TRANSFER STATISTICS
When a print job is transfered to a remote machine (which might be
another unix box, or may be a network printer), it may be useful
to keep statistics on each transfer.  The
.Cm sr
and
.Cm ss
options indicate filenames that lpd should use to store such
statistics.  A statistics line is written for each datafile of a
job as the file is successfully transferred.  The format of the
line is the same for both the sending and receiving side of a
transfer.
.Pp
Statistics on datafiles being received would be used on a print
server, if you are interested in network performance between a
variety of machines which are sending jobs to that print server.
The print server could collect statistics on the speed of each
print job as it arrived on the server.
.Pp
Statistics on datafiles being sent might be used as a minimal
accounting record, when you want to know who sent which jobs to a
remote printer, when they were sent, and how large (in bytes) the
files were.  This will not give include any idea of how many pages
were printed, because there is no standard way to get that information
back from a remote (network) printer in this case.
.Sh LOGGING
Error messages generated by the line printer programs themselves
(that is, the
.Xr lpd 8
and related programs)
are logged by
.Xr syslog 3
using the
.Dv LPR
facility.
Messages printed on
.Em stderr
of one of the filters
are sent to the corresponding
.Cm lf
file.
The filters may, of course, use
.Xr syslog 8
themselves.
.Pp
Error messages sent to the console have a carriage return and a line
feed appended to them, rather than just a line feed.
.Sh SEE ALSO
.Xr lpq 1 ,
.Xr lpr 1 ,
.Xr lprm 1 ,
.Xr hosts.lpd 5 ,
.Xr termcap 5 ,
.Xr lpc 8 ,
.Xr lpd 8 ,
.Xr pac 8
.Rs
.%T "4.3 BSD Line Printer Spooler Manual"
.Re
.Sh HISTORY
The
.Nm
file format appeared in
.Bx 4.2 .


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

freebsd4/usr.sbin/lpr/common_source/printcap.c
/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 * (c) UNIX System Laboratories, Inc.
 * All or some portions of this file are derived from material licensed
 * to the University of California by American Telephone and Telegraph
 * Co. or Unix System Laboratories, Inc. and are reproduced herein with
 * the permission of UNIX System Laboratories, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * 	From: @(#)common.c	8.5 (Berkeley) 4/28/95
 */

#ifndef lint
static const char rcsid[] =
  "$FreeBSD: src/usr.sbin/lpr/common_source/printcap.c,v 1.6 1999/08/28 01:16:48 peter Exp $";
#endif /* not lint */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <sys/param.h>		/* required for lp.h, but not used here */
#include <sys/dirent.h>		/* ditto */
#include "lp.h"
#include "lp.local.h"
#include "pathnames.h"

/*
 * Routines and data used in processing the printcap file.
 */
static 	char *printcapdb[2] = { _PATH_PRINTCAP, 0 }; /* list for cget* */

static	char *capdb_canonical_name(const char *);
static	int capdb_getaltlog(char *, const char *, const char *);
static	int capdb_getaltnum(char *, const char *, const char *, long, long *);
static	int capdb_getaltstr(char *, const char *, const char *, const char *,
			    char **);
static	int getprintcap_int(char *bp, struct printer *pp);

/*
 * Change the name of the printcap file.  Used by chkprintcap(8),
 * but could be used by other members of the suite with appropriate
 * security measures.
 */
void
setprintcap(char *newfile)
{
	printcapdb[0] = newfile;
}

/*
 * Read the printcap database for printer `printer' into the
 * struct printer pointed by `pp'.  Return values are as for
 * cgetent(3): -1 means we could not find what we wanted, -2
 * means a system error occurred (and errno is set), -3 if a
 * reference (`tc=') loop was detected, and 0 means success.
 *
 * Copied from lpr; should add additional capabilities as they
 * are required by the other programs in the suite so that
 * printcap-reading is consistent across the entire family.
 */
int
getprintcap(const char *printer, struct printer *pp)
{
	int status;
	char *XXX;
	char *bp;

	/*
	 * A bug in the declaration of cgetent(3) means that we have
	 * to hide the constness of its third argument.
	 */
	XXX = (char *)printer;
	if ((status = cgetent(&bp, printcapdb, XXX)) < 0)
		return status;
	status = getprintcap_int(bp, pp);
	free(bp);
	return status;
}

/*
 * Map the status values returned by cgetfirst/cgetnext into those
 * used by cgetent, returning truth if there are more records to
 * examine.  This points out what is arguably a bug in the cget*
 * interface (or at least a nasty wart).
 */
static int
firstnextmap(int *status)
{
	switch (*status) {
	case 0:
		return 0;
	case 1:
		*status = 0;
		return 1;
	case 2:
		*status = 1;
		return 1;
	case -1:
		*status = -2;
		return 0;
	case -2:
		*status = -3;
		return 1;
	default:
		return 0;
	}
}

/*
 * Scan through the database of printers using cgetfirst/cgetnext.
 * Return false of error or end-of-database; else true.
 */
int
firstprinter(struct printer *pp, int *error)
{
	int status;
	char *bp;

	init_printer(pp);
	status = cgetfirst(&bp, printcapdb);
	if (firstnextmap(&status) == 0) {
		if (error)
			*error = status;
		return 0;
	}
	if (error)
		*error = status;
	status = getprintcap_int(bp, pp);
	free(bp);
	if (error && status)
		*error = status;
	return 1;
}

int
nextprinter(struct printer *pp, int *error)
{
	int status;
	char *bp;

	free_printer(pp);
	status = cgetnext(&bp, printcapdb);
	if (firstnextmap(&status) == 0) {
		if (error)
			*error = status;
		return 0;
	}
	if (error)
		*error = status;
	status = getprintcap_int(bp, pp);
	free(bp);
	if (error && status)
		*error = status;
	return 1;
}

void
lastprinter(void)
{
	cgetclose();
}

/*
 * This must match the order of declaration of enum filter in lp.h.
 */
static const char *filters[] = {
	"cf", "df", "gf", "if", "nf", "of", "rf", "tf", "vf"
};

static const char *longfilters[] = {
	"filt.cifplot", "filt.dvi", "filt.plot", "filt.input", "filt.ditroff",
	"filt.output", "filt.fortran", "filt.troff", "filt.raster"
};

/*
 * Internal routine for both getprintcap() and nextprinter().
 * Actually parse the printcap entry using cget* functions.
 * Also attempt to figure out the canonical name of the printer
 * and store a malloced copy of it in pp->printer.
 */
static int
getprintcap_int(bp, pp)
	char *bp;
	struct printer *pp;
{
	enum lpd_filters filt;
	char *rp_name;
	int error;

	if ((pp->printer = capdb_canonical_name(bp)) == 0)
		return PCAPERR_OSERR;

#define CHK(x) do {if ((x) == PCAPERR_OSERR) return PCAPERR_OSERR;}while(0)
	CHK(capdb_getaltstr(bp, "af", "acct.file", 0, &pp->acct_file));
	CHK(capdb_getaltnum(bp, "br", "tty.rate", 0, &pp->baud_rate));
	CHK(capdb_getaltnum(bp, "ct", "remote.timeout", DEFTIMEOUT, 
			    &pp->conn_timeout));
	CHK(capdb_getaltnum(bp, "du", "daemon.user", DEFUID, 
			    &pp->daemon_user));
	CHK(capdb_getaltstr(bp, "ff", "job.formfeed", DEFFF, &pp->form_feed));
	CHK(capdb_getaltstr(bp, "lf", "spool.log", _PATH_CONSOLE, 
			    &pp->log_file));
	CHK(capdb_getaltstr(bp, "lo", "spool.lock", DEFLOCK, &pp->lock_file));
	CHK(capdb_getaltstr(bp, "lp", "tty.device", _PATH_DEFDEVLP, &pp->lp));
	CHK(capdb_getaltnum(bp, "mc", "max.copies", DEFMAXCOPIES, 
			    &pp->max_copies));
	CHK(capdb_getaltstr(bp, "ms", "tty.mode", 0, &pp->mode_set));
	CHK(capdb_getaltnum(bp, "mx", "max.blocks", DEFMX, &pp->max_blocks));
	CHK(capdb_getaltnum(bp, "pc", "acct.price", 0, &pp->price100));
	CHK(capdb_getaltnum(bp, "pl", "page.length", DEFLENGTH,
			    &pp->page_length));
	CHK(capdb_getaltnum(bp, "pw", "page.width", DEFWIDTH, 
			    &pp->page_width));
	CHK(capdb_getaltnum(bp, "px", "page.pwidth", 0, &pp->page_pwidth));
	CHK(capdb_getaltnum(bp, "py", "page.plength", 0, &pp->page_plength));
	CHK(capdb_getaltstr(bp, "rg", "daemon.restrictgrp", 0, 
			    &pp->restrict_grp));
	CHK(capdb_getaltstr(bp, "rm", "remote.host", 0, &pp->remote_host));
	CHK(capdb_getaltstr(bp, "rp", "remote.queue", DEFLP, 
			    &pp->remote_queue));
	CHK(capdb_getaltstr(bp, "sd", "spool.dir", _PATH_DEFSPOOL,
			    &pp->spool_dir));
	CHK(capdb_getaltstr(bp, "sr", "stat.recv", 0, &pp->stat_recv));
	CHK(capdb_getaltstr(bp, "ss", "stat.send", 0, &pp->stat_send));
	CHK(capdb_getaltstr(bp, "st", "spool.status", DEFSTAT,
			    &pp->status_file));
	CHK(capdb_getaltstr(bp, "tr", "job.trailer", 0, &pp->trailer));

	pp->restricted = capdb_getaltlog(bp, "rs", "daemon.restricted");
	pp->short_banner = capdb_getaltlog(bp, "sb", "banner.short");
	pp->no_copies = capdb_getaltlog(bp, "sc", "job.no_copies");
	pp->no_formfeed = capdb_getaltlog(bp, "sf", "job.no_formfeed");
	pp->no_header = capdb_getaltlog(bp, "sh", "banner.disable");
	pp->header_last = capdb_getaltlog(bp, "hl", "banner.last");
	pp->rw = capdb_getaltlog(bp, "rw", "tty.rw");
	pp->tof = !capdb_getaltlog(bp, "fo", "job.topofform");
	
	/*
	 * Decide if the remote printer name matches the local printer name.
	 * If no name is given then we assume they mean them to match.
	 * If a name is given see if the rp_name is one of the names for
	 * this printer.
	 */
	pp->rp_matches_local = 1;
	CHK((error = capdb_getaltstr(bp, "rp", "remote.queue", 0, &rp_name)));
	if (error != PCAPERR_NOTFOUND && rp_name != NULL) {
		if (cgetmatch(bp,rp_name) != 0)
			pp->rp_matches_local = 0;
		free(rp_name);
	}

	/*
	 * Filters:
	 */
	for (filt = 0; filt < LPF_COUNT; filt++) {
		CHK(capdb_getaltstr(bp, filters[filt], longfilters[filt], 0,
				    &pp->filters[filt]));
	}

	return 0;
}

/*
 * Decode the error codes returned by cgetent() using the names we
 * made up for them from "lp.h".
 * This would have been much better done with Common Error, >sigh<.
 * Perhaps this can be fixed in the next incarnation of cget*.
 */
const char *
pcaperr(int error)
{
	switch(error) {
	case PCAPERR_TCOPEN:
		return "unresolved tc= expansion";
	case PCAPERR_SUCCESS:
		return "no error";
	case PCAPERR_NOTFOUND:
		return "printer not found";
	case PCAPERR_OSERR:
		return strerror(errno);
	case PCAPERR_TCLOOP:
		return "loop detected in tc= expansion";
	default:
		return "unknown printcap error";
	}
}

/*
 * Initialize a `struct printer' to contain values harmless to
 * the other routines in liblpr.
 */
void
init_printer(struct printer *pp)
{
	static struct printer zero;
	*pp = zero;
}

/*
 * Free the dynamically-allocated strings in a `struct printer'.
 * Idempotent.
 */
void
free_printer(struct printer *pp)
{
	enum lpd_filters filt;
#define	cfree(x)	do { if (x) free(x); } while(0)
	cfree(pp->printer);
	cfree(pp->acct_file);
	for (filt = 0; filt < LPF_COUNT; filt++)
		cfree(pp->filters[filt]);
	cfree(pp->form_feed);
	cfree(pp->log_file);
	cfree(pp->lock_file);
	cfree(pp->lp);
	cfree(pp->restrict_grp);
	cfree(pp->remote_host);
	cfree(pp->remote_queue);
	cfree(pp->spool_dir);
	cfree(pp->stat_recv);
	cfree(pp->stat_send);
	cfree(pp->status_file);
	cfree(pp->trailer);
	cfree(pp->mode_set);

	init_printer(pp);
}


/* 
 * The following routines are part of what would be a sensible library 
 * interface to capability databases.  Maybe someday this will become
 * the default.
 */

/*
 * It provides similar functionality to cgetstr(),
 * except that it provides for both a long and a short
 * capability name and allows for a default to be specified.
 */
static int
capdb_getaltstr(char *bp, const char *shrt, const char *lng, 
		const char *dflt, char **result)
{
	int status;

	status = cgetstr(bp, (char *)/*XXX*/lng, result);
	if (status >= 0 || status == PCAPERR_OSERR)
		return status;
	status = cgetstr(bp, (char *)/*XXX*/shrt, result);
	if (status >= 0 || status == PCAPERR_OSERR)
		return status;
	if (dflt) {
		*result = strdup(dflt);
		if (*result == 0)
			return PCAPERR_OSERR;
		return strlen(*result);
	}
	return PCAPERR_NOTFOUND;
}

/*
 * The same, only for integers.
 */
static int
capdb_getaltnum(char *bp, const char *shrt, const char *lng, long dflt,
		long *result)
{
	int status;

	status = cgetnum(bp, (char *)/*XXX*/lng, result);
	if (status >= 0)
		return status;
	status = cgetnum(bp, (char *)/*XXX*/shrt, result);
	if (status >= 0)
		return status;
	*result = dflt;
	return 0;
}	

/*
 * Likewise for logical values.  There's no need for a default parameter
 * because the default is always false.
 */
static int
capdb_getaltlog(char *bp, const char *shrt, const char *lng)
{
	if (cgetcap(bp, (char *)/*XXX*/lng, ':'))
		return 1;
	if (cgetcap(bp, (char *)/*XXX*/shrt, ':'))
		return 1;
	return 0;
}

/*
 * Also should be a part of a better cget* library.
 * Given a capdb entry, attempt to figure out what its canonical name
 * is, and return a malloced copy of it.  The canonical name is
 * considered to be the first one listed.
 */
static char *
capdb_canonical_name(const char *bp)
{
	char *retval;	
	const char *nameend;

	nameend = strpbrk(bp, "|:");
	if (nameend == 0)
		nameend = bp + 1;
	if ((retval = malloc(nameend - bp + 1)) != 0) {
		retval[0] = '\0';
		strncat(retval, bp, nameend - bp);
	}
	return retval;
}


